File: src/main/java/com/grepp/curdsample/domain/Task.java
package com.grepp.curdsample.domain;

import com.grepp.curdsample.dto.TaskDto;
import com.grepp.curdsample.util.TimeFormatter;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Getter
@Entity
@Table(name = "tasks")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Task {

    @Id
//    @Column(name = "tasks_id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Setter
    @Column(nullable = false, unique = true)
    private String code;

    private String title;
    private String description;

    private Integer priority;

//    @Setter
    private boolean completeStatus = false;

    private LocalDate startTime;
    private LocalDate endTime;

    private LocalDateTime createdAt = LocalDateTime.now();
    private LocalDateTime updatedAt = LocalDateTime.now();

    public void updateCheck() {
        this.completeStatus = !this.completeStatus;
    }

    public void update(TaskDto dto) {

        this.title = dto.getTitle();
        this.description = dto.getDescription();
        this.priority = dto.getPriority();
        this.completeStatus = dto.isCompleteStatus();

        this.startTime = TimeFormatter.convertToLocalDate(dto.getStartTime());
        this.endTime = TimeFormatter.convertToLocalDate(dto.getEndTime());

    }

    public static Task of(TaskDto taskDto) {
        Task task = new Task();

        task.code = taskDto.getCode();
        task.title = taskDto.getTitle();
        task.description = taskDto.getDescription();
        task.priority = taskDto.getPriority();
        task.completeStatus = taskDto.isCompleteStatus();
        task.startTime = TimeFormatter.convertToLocalDate(taskDto.getStartTime());
        task.endTime = TimeFormatter.convertToLocalDate(taskDto.getEndTime());

        return task;
    }


}

File: src/main/java/com/est/curdsample/domain/Task.kt
package com.est.curdsample.domain

import com.est.curdsample.dto.TaskDescription
import com.est.curdsample.dto.TaskDto
import com.est.curdsample.util.TimeFormatter
import jakarta.persistence.*
import lombok.Setter
import java.time.LocalDate
import java.time.LocalDateTime


@Entity
@Table(name = "tasks")
class Task(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @Setter
    @Column(nullable = false, unique = true)
    val code: String,

    var title: String,

    var description: String,

    var priority: Int,

    var completeStatus: Boolean,

    var startTime: LocalDate,

    var endTime: LocalDate,

    val createdAt: LocalDateTime = LocalDateTime.now(),

    var updatedAt: LocalDateTime = LocalDateTime.now()
) {

    fun updateCheck() {
        this.completeStatus = !this.completeStatus
        this.updatedAt = LocalDateTime.now()
    }

    fun update(dto: TaskDto) {
        this.title = dto.title
        this.description = dto.description
        this.priority = dto.priority
        this.completeStatus = dto.completeStatus

        this.startTime = TimeFormatter.convertToLocalDate(dto.startTime)
        this.endTime = TimeFormatter.convertToLocalDate(dto.endTime)
        this.updatedAt = LocalDateTime.now()
    }
}

fun Task.toDto() : TaskDto {
    return TaskDto(
        code = this.code,
        title = this.title,
        description = this.description,
        priority = this.priority,
        completeStatus = this.completeStatus,
        startTime = TimeFormatter.convertToStr(this.startTime),
        endTime = TimeFormatter.convertToStr(this.endTime),
    )
}

fun Task.toDescription() : TaskDescription {
    return TaskDescription(
        code = this.code,
        title = this.title,
        description = this.description,
        priority = this.priority,
        completeStatus = this.completeStatus,
        startDate = TimeFormatter.convertToStr(this.startTime),
        dueDate = TimeFormatter.convertToStr(this.endTime),
        createdAt = this.createdAt,
        updatedAt = this.updatedAt
    )
}